mixer/pkg/protobuf/yaml/decoder.go

```diff
@@ -223,10 +237,54 @@ func (dv *decodeVisitor) Bytes(n wire.Number, v []byte) {
 				v = v[m:]
 			}
 
-			dv.out[f.GetName()].(map[string]string)[visitor.key] = visitor.value
+			m[visitor.key] = visitor.value
 			return
 		}
-		// TODO(kuat): implement sub-message decoding
+
+		// parse the entirety of the message recursively for value types
+		if _, ok := valueResolver.descriptors[f.GetTypeName()]; ok && !f.IsRepeated() {
+			decoder := NewDecoder(valueResolver, f.GetTypeName(), nil, "")
+			inner := attribute.GetMutableBag(nil)
+			defer inner.Done()
+			if err := decoder.Decode(v, inner); err != nil {
+				dv.err = multierror.Append(dv.err, fmt.Errorf("failed to decode field %q: %v", f.GetName(), err))
+				return
+			}
+
+			switch f.GetTypeName() {
+			case ".istio.policy.v1beta1.Value":
+				// pop oneof inner value from the bag to the field value
+				names := inner.Names()
+				if len(names) == 1 {
+					if val, ok := inner.Get(names[0]); ok {
+						dv.out.Set(name, val)
+					}
+				}
+			case ".istio.policy.v1beta1.IPAddress",
+				".istio.policy.v1beta1.Duration",
+				".istio.policy.v1beta1.DNSName":
```

Comment:
just curious: does this cover the full set of types that can be returned?  I can't recall.

Reply:
The complete list is here: https://github.com/istio/api/blob/master/policy/v1beta1/value_type.proto . Some things are not there (e.g. `EMAIL_ADDRESS`) because there are nowhere to be found in the code base. I think it's easy to extend this list as the need arises.

Reply:
Can we add a note / TODO here to track this, just in case? I worry that this _could_ lead to a situation in which we half-support some `ValueType` down-the-road.

Is `Timestamp` needed here yet?

Reply:
I'll file an issue and can follow-up. I haven't seen an example for time stamp, and it has limited coverage across the code base AFAIK. The status for EMAIL_ADDRESS is unclear to me, since it is marked as experimental in the code, but yet is part of API.

Reply:
https://github.com/istio/istio/issues/9415